--- libgcc/Makefile.in.orig	2012-09-21 14:24:18.902021415 -0300
+++ libgcc/Makefile.in	2012-09-21 14:28:32.976443070 -0300
@@ -500,14 +500,13 @@
 lib2-divmod-o = $(patsubst %,%$(objext),$(LIB2_DIVMOD_FUNCS))
 $(lib2-divmod-o): %$(objext): $(srcdir)/libgcc2.c
 	$(gcc_compile) -DL$* -c $< \
-	  -fexceptions -fnon-call-exceptions $(vis_hide)
+	  $(vis_hide)
 libgcc-objects += $(lib2-divmod-o)
 
 ifeq ($(enable_shared),yes)
 lib2-divmod-s-o = $(patsubst %,%_s$(objext),$(LIB2_DIVMOD_FUNCS))
 $(lib2-divmod-s-o): %_s$(objext): $(srcdir)/libgcc2.c
-	$(gcc_s_compile) -DL$* -c $< \
-	  -fexceptions -fnon-call-exceptions
+	$(gcc_s_compile) -DL$* -c $<
 libgcc-s-objects += $(lib2-divmod-s-o)
 endif
 
