--- gcc/gengtype.c.orig	2012-12-11 00:42:29.000000000 -0800
+++ gcc/gengtype.c	2012-12-31 14:53:26.000000000 -0800
@@ -3651,13 +3651,13 @@
 		  int has_length, struct fileloc *line, const char *if_marked,
 		  bool emit_pch, type_p field_type, const char *field_name)
 {
-  struct pair newv;
   /* If the field reference is relative to V, rather than to some
      subcomponent of V, we can mark any subarrays with a single stride.
      We're effectively treating the field as a global variable in its
      own right.  */
   if (v && type == v->type)
     {
+      struct pair newv;
       newv = *v;
       newv.type = field_type;
       newv.name = ACONCAT ((v->name, ".", field_name, NULL));
