--- gcc/gengtype.c	2012-09-07 18:39:22.550073229 -0300
+++ gcc/gengtype.c	2012-09-07 18:39:39.645963181 -0300
@@ -3598,9 +3598,9 @@
      subcomponent of V, we can mark any subarrays with a single stride.
      We're effectively treating the field as a global variable in its
      own right.  */
+  struct pair newv;
   if (v && type == v->type)
     {
-      struct pair newv;
 
       newv = *v;
       newv.type = field_type;
